lib/eventHubSender.ts

```diff
@@ -0,0 +1,335 @@
+      for (let i = 0; i < datas.length; i++) {
+        let message = EventData.toAmqpMessage(datas[i]);
+        if (partitionKey) {
+          if (!message.message_annotations) message.message_annotations = {};
+          message.message_annotations[Constants.partitionKey] = partitionKey;
+        }
+        messages[i] = message;
+      }
+      // Encode every amqp message and then convert every encoded message to amqp data section
+      let batchMessage: AmqpMessage = {
+        body: rhea.message.data_sections(messages.map(rhea.message.encode))
+      };
+      // Set message_annotations, application_properties and properties of the first message as
+      // that of the envelope (batch message).
+      if (messages[0].message_annotations) {
+        batchMessage.message_annotations = messages[0].message_annotations;
+      }
+      if (messages[0].application_properties) {
+        batchMessage.application_properties = messages[0].application_properties;
+      }
+      if (messages[0].properties) {
+        batchMessage.properties = messages[0].properties;
+      }
+      // Finally encode the envelope (batch message).
+      const encodedBatchMessage = rhea.message.encode(batchMessage);
+      debug(`[${this._context.connectionId}] Sender "${this.name}", ` +
+        `sending encoded batch message.`, encodedBatchMessage);
+      return await this._trySend(encodedBatchMessage, undefined, 0x80013700);
+    } catch (err) {
+      return Promise.reject(err);
+    }
+  }
+
+  /**
+   * "Unlink" this sender, closing the link and resolving when that operation is complete.
+   * Leaves the underlying connection/session open.
+   * @method close
+   * @return {Promise<void>} Promise<void>
+   */
+  async close(): Promise<void> {
+    if (this._sender) {
+      try {
+        await rheaPromise.closeSender(this._sender);
+        this.removeAllListeners();
+        delete this._context.senders[this.name!];
+        debug(`Deleted the sender "${this.name!}" from the client cache.`);
+        this._sender = undefined;
+        this._session = undefined;
+        clearTimeout(this._tokenRenewalTimer as NodeJS.Timer);
+        debug(`[${this._context.connectionId}] Sender "${this.name}" closed.`);
+      } catch (err) {
+        return Promise.reject(err);
+      }
+    }
+  }
+
+  /**
+   * Tries to send the message to EventHub if there is enough credit to send them
+   * and the circular buffer has available space to settle the message after sending them.
+   *
+   * We have implemented a synchronous send over here. We shall be waiting for the message
+   * to be accepted or rejected and accordingly resolve or reject the promise.
+   *
+   * @param message The message to be sent to EventHub.
+   * @return {Promise<rheaPromise.Delivery>} Promise<rheaPromise.Delivery>
+   */
+  private _trySend(message: AmqpMessage, tag?: any, format?: number): Promise<rheaPromise.Delivery> {
+    return new Promise((resolve, reject) => {
```

Comment:
For now I have implemented synchronous send in the sense that I send a message and then wait for the accepted/rejected event to occur. Then I resolve/reject the promise accordingly.

Reply:
this is what i would expect. the send isn't really synchronous so much as it allows you to perform sends sequentially (if desired)